### PR TITLE
Add config validation

### PR DESCRIPTION
**Pull request by PeteBa**
**Describe the change**
An early view at adding validation of user's configuration file to see if this is worthwhile and heading in the right direction.

This leverages the `class-validator` ([link](https://github.com/typestack/class-validator)) that provides typescript decorators for validation rules and hooks into `node-validator` for ,,, validation. This is only a proof-of-concept for generating some feedback/direction rather than polishing reviews. The code wont take a robust review at this point 8)

The library looks OK and has a reasonable history/activity going on. I want to take a look at joi/joiful-ts ([link](https://github.com/joiful-ts/joiful)) to get another perspective particularly with joi, having a bit more weight behind it. But open to exploring other options here as to be honest the work is more in the schema/rules than the code plumbing.

Should be OK to spin up and try running your local.yml file through the validation to get an idea of how it works. But as mentioned, the validation is POC and specifically nested config e.g. entity/behaviours etc are not yet implemented. The error reporting is there but just leveraging out-of-box formatting which could be improved. Doesn't support file/line identification for errors rather relies on Integration/Property to steer the user. Some of the room-assistant config is moderately complex and needs some further work to find the right validators.

Appreciate any feedback and let me know if this needs a re-point as happy to revise as needed. Tests and Docs will come down the line.

**Checklist**
If you changed code:
- [ ] Tests run locally and pass (`npm test`)
- [ ] Code has correct format (`npm run format`)

If you added a new integration:
- [ ] Documentation page added in `docs/integrations/`
- [ ] Page linked in `docs/.vuepress/config.js` and `docs/integrations/README.md`

**Additional information**
Is this PR related to any issues? Do you have anything else to add?

### CODE REVIEW

**Comment by mKeRix** [comment]
Color me impressed! From a first high-level view this looks really good.

I have no particular opinions on the tooling, both suggestions seem fine to me as of right now. I think `class-validator` has the benefit of already being a dependency for room-assistant because of the OpenAPI stuff and being a bit more active. `joiful-ts` claims to allow for slightly more complex validation though, which makes it an interesting contender. I'd trust your choice out of those two.

I don't think I'd want to change anything about the general direction you took here. As a remark though: I think the config validation should not cause the application to shut down, at least not yet. Logging it as warn or error, like you've chosen to do right now, seems like the right approach to me.

**Comment by PeteBa** [comment]
Moved this over to `Joiful` to support needed validations and access to full Joi API. Benefit is nicer decorators (chaining) and error messaging. Tricky choice as `class-validator` is very good and active community. My thinking is that Joiful is mostly a thin layer on top of Joi which has a very substantive base. Happy to review.

I have attached a near full-coverage `local.yml` file that I have used for informal test that contains a bunch of errors and generates the output below by way of example. Be good to get additions/corrections as some of the validators are my intepretation (e.g. number vs integer). Will think about how best to incorporate into jest in the next commit.

I think there are a couple of other things to look at: 1) abstracting `config` (i.e. import c from 'config') into `ConfigServices` module, and; 2) better sequence config load and validation to avoid pre-use. Neither of these is a dependency on getting this buttoned down.

I agree with the "non-fatal" strategy until there has been adequate road-testing. A lot of parameters and undoubtable some gotchas in there.

[local.yml.txt](https://github.com/mKeRix/room-assistant/files/6064002/local.yml.txt)

```
3/1/2021, 19:03:28 - info - ConfigService: Loading configuration from /home/peter/room-assistant/src/config/definitions/default.ts, config/local.yml (Current: /home/peter/room-assistant)
3/1/2021, 19:03:28 - error - ConfigService: "logger.elasticsearch.enabled" must be a boolean [Value: 123]
3/1/2021, 19:03:28 - error - ConfigService: "cluster.timeout" must be greater than or equal to 0 [Value: -3]
3/1/2021, 19:03:28 - error - ConfigService: "entities.behaviors.debounced_entity.debounce.leading" is required [Value: undefined]
3/1/2021, 19:03:28 - error - ConfigService: "entities.behaviors.chained_entity.debounce.leading" is required [Value: undefined]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothLowEnergy.minorMask" must be a number [Value: "0xGHIG"]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothLowEnergy.tagOverrides.ebef1234567890-55555-333.measuredPower" must be a negative number [Value: 33]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothLowEnergy.tagOverrides.ebef1234567890-55555-444.fake" is not allowed [Value: "fake"]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothLowEnergy.updateFrequence" is not allowed [Value: 55]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.addresses[1]" does not match the required MAC address format [Value: "77:50:fb:4d:ab:7G"]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.addresses[2]" does not match the required MAC address format [Value: "77:50"]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.addresses[3]" does not match the required MAC address format [Value: "77::fb:4d:ab:70"]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.minRssi" does not match any of the allowed types [Value: {"08:05:90:ed:3b:60":10,"08:05:90:ed:32:1Z":-10,"default":-20}]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.timeoutCycles" must be an integer [Value: -2.5]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.timeoutCycles" must be greater than or equal to 0 [Value: -2.5]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.entityOverrides.ebef1234567890-55555-333.id" must be a string [Value: 333]
3/1/2021, 19:03:28 - error - ConfigService: "bluetoothClassic.hciiDeviceId" is not allowed [Value: 1]
3/1/2021, 19:03:28 - error - ConfigService: "omronD6t.heatmap.enabled" must be a boolean [Value: 123]
3/1/2021, 19:03:28 - error - ConfigService: "gridEye.heatmap.minTemperature" must be a number [Value: "23"]
3/1/2021, 19:03:28 - error - ConfigService: "gridEye.heatmap.rotation" must be one of [0, 90, 180, 270] [Value: 179]
3/1/2021, 19:03:28 - error - ConfigService: "gpio.binarySensors[1].deviceClass" must be one of [battery, cold, connectivity, door, garage_door, gas, heat, light, lock, moisture, motion, moving, occupancy, opening, plug, power, presence, problem, safety, smoke, sound, vibration, window] [Value: "fake-device-class"]
3/1/2021, 19:03:28 - error - ConfigService: "xiaomiMi.sensors[0].address" must only contain hexadecimal characters [Value: "a4:c1:38:32:12:f9"]
3/1/2021, 19:03:28 - error - ConfigService: "xiaomiMi.sensors[0].address" length must be 12 characters long [Value: "a4:c1:38:32:12:f9"]
3/1/2021, 19:03:28 - error - ConfigService: "xiaomiMi.sensors[1].bindKey" must only contain hexadecimal characters [Value: "2sddssfsdf657568sadfs 78"]
3/1/2021, 19:03:28 - error - ConfigService: "homeAssistant.fake-property" is not allowed [Value: "fasfasfda"]
3/1/2021, 19:03:28 - error - ConfigService: "fake-integration" is not allowed [Value: null]
```

**Comment by mKeRix** [comment]
I think those are very valid reasons to choose Joiful. Again, I'm very impressed by the effort you've put into this. Lots of good stuff. :)

I'm happy to do a re-review and looking at all the validations you put in place in more detail whenever you request it. Some things might only become apparent with real world usage, but I think that's fine too. The room-assistant config has become a bit complex in some spots, so it will be hard to really think of everything.

I like your thoughts on load order and abstracting away the direct config calls, but it's no showstopper for me either. Some of it might be a bit tricky to resolve, since I used the direct calls for bending some stuff to fit into the framework of the project.

**Comment by PeteBa** [comment]
Should be fine now to review the validations as I'm now looking at test approach. Be interested to know the "top 5" config errors and whether caught, e.g. missing/misspelled keys, mac address format, incorrect nesting, ...

For tests, I'm thinking three things: 1) a "pass" file containing a full  and valid config used to highlight regressions, 2) unit tests for custom validation functions: 3) any specific edge-cases (?). Want to avoid testing "joiful/joi" itself.

File: src/integrations/bluetooth-low-energy/bluetooth-low-energy.config.ts [hunk 1]
```diff
@@ -1,28 +1,67 @@
...
   maxDistance?: number;
 }
+// TODO OR Properties(whitelist|allowlist)
 
-class TagOverride {
-  id?: string;
-  name?: string;
-  measuredPower?: number;
-  batteryMask?: number;
+const BluetoothLowEnergyScheme = jf
```
Comment (mKeRix) [inline src/integrations/bluetooth-low-energy/bluetooth-low-energy.config.ts:55]: Is this used somehow or just left over? If it is used by some joi magic then I think it would be a good idea to suppress the linter warning for this line, otherwise it can be removed imo.

**Review by mKeRix — CHANGES_REQUESTED** [review]
I mostly had minor remarks, your overall interpretation is quite close to the reality I'd say.

As for the tests, I think your plans are fine. Testing one valid and one invalid configuration could also be enough. I don't think you need to invest a lot of effort into the testing part, since all the annotations are just descriptions that don't need to be tested per se imo. It would just be nice to have a test that verifies that the joi validations are triggered and output errors somewhere.

File: src/integrations/home-assistant/home-assistant.config.ts [hunk 2]
```diff
@@ -1,8 +1,13 @@
 import { IClientOptions } from 'async-mqtt';
+import * as jf from 'joiful';
 
 export class HomeAssistantConfig {
+  @(jf.string().required())
```
[the remark this answers is not in the document]
Comment (PeteBa) [reply src/integrations/home-assistant/home-assistant.config.ts]: Nice

**Comment by PeteBa** [comment]
Rebased and test coverage added. Test is three fold: a "pass" file for for coverage and regression; a "fail" file that covers main use cases; specific unit tests for custom code. The "fail" file feels a bit heavy and dupes much of the pass file content but provides an easy way to add test quick. Could move to a diff on the pass file down the line.

I have one known snag at `config.service.ts L47` with a lint override. Need a fresh brain to bottom it  out!

File: src/config/config.service.ts [hunk 3]
```diff
@@ -27,6 +32,34 @@ export class ConfigService implements OnModuleInit {
...
+  /**
+   * Validates the union of default and user configuration parameters. The parser will
+   * not abort on first error but will highlight all errors detected. Validation is done
+   * strictly without conversion of types. An exception is raised on validation failure.
+   *
+   * @param cfg - configuration object to validate
+   */
+  // TODO Remove need for lint skip - need inspiration from the type gods
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  validateConfig(cfg: Object): void {
```
Comment (mKeRix) [inline src/config/config.service.ts:48]: I think one way to not use the object type would be using the `IConfig` type from the `config` package - just as an idea

Comment (PeteBa) [reply src/config/config.service.ts]: Just seen this ... I swear I tried every combination 8) Thanks.

**Comment by PeteBa** [comment]
OK. I am done. Let me know any feedback.

**Review by mKeRix — APPROVED** [review]
Looks good! Thanks for all your work, LGTM :)
